### PR TITLE
actions: setup image and Helm chart publishing

### PR DESCRIPTION
**Pull request by taraspos**
### Summary

Configure CI/CD build for docker images and helm charts.

Mostly copied from [aws-quota-checker](https://github.com/gravitational/aws-quota-checker/blob/teleport/.github/workflows/cd.yml) repo with small modifications.

Every commit to `main` is released to private repo, while only tags are released public repo.

### Testing

Not tested, need to merge to main first to be able to dispatch the workflow

### CODE REVIEW

File: .github/workflows/cd.yaml [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+      - name: Configure AWS credentials for ECR Private
+        uses: aws-actions/configure-aws-credentials@v4
+        with:
+          aws-region: ${{ env.AWS_REGION_PRIVATE }}
+          role-to-assume: ${{ env.ECR_ROLE }}
+          mask-aws-account-id: 'no'
+
+      - name: Configure AWS credentials for ECR Public
+        if: "startsWith(github.ref, 'refs/tags/v')"
+        uses: aws-actions/configure-aws-credentials@v4
```
Comment (evanfreed) [inline .github/workflows/cd.yaml]:
```suggestion
        uses: aws-actions/configure-aws-credentials@e3dd6a429d7300a6a4c196c26e071d42e0343502 # v4
```

nit: prefer hash over tag

File: .github/workflows/cd.yaml [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      - name: Checkout repo
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Setup docker buildx
+        uses: docker/setup-buildx-action@f95db51fddba0c2d1ec667646a06c2ce06100226 # v3.0.0
+        
+      - name: Configure AWS credentials for ECR Private
+        uses: aws-actions/configure-aws-credentials@v4
```
Comment (evanfreed) [inline .github/workflows/cd.yaml]:
```suggestion
        uses: aws-actions/configure-aws-credentials@e3dd6a429d7300a6a4c196c26e071d42e0343502 # v4
```

File: .github/workflows/cd.yaml [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+          role-to-assume: ${{ env.ECR_ROLE }}
+          mask-aws-account-id: 'no'
+
+      - name: Configure AWS credentials for ECR Public
+        if: "startsWith(github.ref, 'refs/tags/v')"
+        uses: aws-actions/configure-aws-credentials@v4
+        with:
+          aws-region: ${{ env.AWS_REGION_PUBLIC }}
+          role-to-assume: ${{ env.ECR_ROLE }}
+          mask-aws-account-id: 'no'
```
Comment (evanfreed) [inline .github/workflows/cd.yaml]:
```suggestion
```
I think the default is no so could be excluded

File: .github/workflows/cd.yaml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+      - name: Setup docker buildx
+        uses: docker/setup-buildx-action@f95db51fddba0c2d1ec667646a06c2ce06100226 # v3.0.0
+        
+      - name: Configure AWS credentials for ECR Private
+        uses: aws-actions/configure-aws-credentials@v4
+        with:
+          aws-region: ${{ env.AWS_REGION_PRIVATE }}
+          role-to-assume: ${{ env.ECR_ROLE }}
+          mask-aws-account-id: 'no'
```
Comment (evanfreed) [inline .github/workflows/cd.yaml]:
```suggestion
```
nit: I think the default is no so could be excluded

File: .github/workflows/cd.yaml [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+      - name: Login to Amazon ECR Private
+        id: login-ecr
+        uses: aws-actions/amazon-ecr-login@v2
+
+      - name: Login to Amazon ECR Public
+        if: "startsWith(github.ref, 'refs/tags/v')"
+        id: login-ecr-public
+        uses: aws-actions/amazon-ecr-login@v2
```
Comment (evanfreed) [inline .github/workflows/cd.yaml]:
```suggestion
        uses: aws-actions/amazon-ecr-login@062b18b96a7aff071d4dc91bc00c4c1a7945b076 # v2
```

File: .github/workflows/cd.yaml [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        if: "startsWith(github.ref, 'refs/tags/v')"
+        uses: aws-actions/configure-aws-credentials@v4
+        with:
+          aws-region: ${{ env.AWS_REGION_PUBLIC }}
+          role-to-assume: ${{ env.ECR_ROLE }}
+          mask-aws-account-id: 'no'
+
+      - name: Login to Amazon ECR Private
+        id: login-ecr
+        uses: aws-actions/amazon-ecr-login@v2
```
Comment (evanfreed) [inline .github/workflows/cd.yaml]:
```suggestion
        uses: aws-actions/amazon-ecr-login@062b18b96a7aff071d4dc91bc00c4c1a7945b076 # v2
```

**Review by evanfreed — APPROVED** [review]
overall looks good, few small nits

File: .github/workflows/cd.yaml [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+permissions:
+  contents: read
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@v4
```
Comment (lcharkiewicz) [inline .github/workflows/cd.yaml]: Similarly to other places: let's use SHA

File: .github/workflows/cd.yaml [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    environment: release
+    env:
+      AWS_REGION_PRIVATE: us-west-2
+      AWS_REGION_PUBLIC: us-east-1
+    permissions:
+      contents: read
+      id-token: write # This is required for requesting the JWT, see https://github.com/aws-actions/configure-aws-credentials#OIDC
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@v4
```
Comment (lcharkiewicz) [inline .github/workflows/cd.yaml]: Ditto.

File: .github/workflows/cd.yaml [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+      - name: Login to Amazon ECR Public
+        if: "startsWith(github.ref, 'refs/tags/v')"
+        id: login-ecr-public
+        uses: aws-actions/amazon-ecr-login@062b18b96a7aff071d4dc91bc00c4c1a7945b076 # v2
+        with:
+          registry-type: public
+
+      - name: Prepare docker labels and tags
+        id: meta
+        uses: docker/metadata-action@v5
```
Comment (lcharkiewicz) [inline .github/workflows/cd.yaml]: And one more.

File: .github/workflows/cd.yaml [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+          # Enable sha tag on branch push events and workflow dispatches.
+          # Enable semver tags on tag push events, but don't overwrite major/minor tags for prereleases.
+          tags: |
+            type=sha,prefix={{branch}}-,suffix=-{{date 'YYYYMMDDTHHmmss'}},format=short,enable=${{ startsWith(github.ref, 'refs/heads/') }}
+            type=semver,pattern={{major}},enable=${{ github.event_name == 'push' && startsWith(github.ref, 'refs/tags/v') && !contains(github.ref, '-') }}
+            type=semver,pattern={{major}}.{{minor}},enable=${{ github.event_name == 'push' && startsWith(github.ref, 'refs/tags/v') && !contains(github.ref, '-') }}
+            type=semver,pattern={{version}},enable=${{ github.event_name == 'push' && startsWith(github.ref, 'refs/tags/v') }}
+  
+      - name: Build the Docker image and push
+        uses: docker/build-push-action@v5
```
Comment (lcharkiewicz) [inline .github/workflows/cd.yaml]: Aaand the last one. :)

**Review by lcharkiewicz — APPROVED** [review]
LGTM, a few nits regarding version SHA pinning.

**Pull request merged**